### DISCUSSION TITLE
List S3 objects in Juice client

### DISCUSSION THREAD

**Discussion by zhanghongzhou** [juicedata/juicefs#4077] [General]
**What would you like to be added**:

像S3FS、Goofys一样，在readdir 目录的时候，把juicefs的该目录的文件，和S3的对应前缀的object list，合并展示出来

**Why is this needed**:
全量数据的glob

**Comment by SandyXSD** [comment]
In JuiceFS, files are split to multiple objects ([doc](https://juicefs.com/docs/community/architecture/#how-juicefs-store-files)), so it makes no sense to show object list in readdir. You can inspect all objects related to a file with the `juicefs info` command.

**Comment by zhanghongzhou** [comment]
这里有个常见场景：海量的存量数据在S3，预热的部分（真子集）在juicefs。用户想全量glob出来，区分哪些需要预热到juicefs

从juicefs写入S3的object在listobject的时候过滤掉，展示给客户端的是从juicefs写入的文件+从S3写入的存量文件

**Comment by davies** [comment]
感觉这个应该属于业务侧的事情，不应该在文件系统里面实现，否则会导致各种不可修复的性能和一致性问题。

**Reply by zhanghongzhou** [reply]
可以做个开关。没有对S3存量数据的只读能力（list、read）的话，存量数据需要预热到juicefs，这个过程很浪费带宽和时间

**Reply by zhanghongzhou** [reply]
一致性应该不是问题，这个需求可以接受最终一致

**Reply by davies** [reply]
JuiceFS 是提供一致性保障的，还是建议业务层自己做下合并。

**Reply by zhanghongzhou** [reply]
看了下alluxio是支持自动加载的：https://docs.alluxio.io/os/user/stable/cn/core-services/Unified-Namespace.html

**Reply by davies** [reply]
JuiceFS 跟 Alluxio 是完全不同的架构，没办法兼顾的。如果 Alluxio 更适合你们的场景，建议直接使用它。
